### PR TITLE
FilteredTableModel

### PR DESCRIPTION
**Pull request by shaobinspire**

### CODE REVIEW

**Review by spireken — APPROVED** [review]
Looks good.

Will factory methods be added for standard filter functions?

File: Applications/Spire/Include/Spire/Ui/FilteredTableModel.hpp [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+#include <functional>
+#include <tuple>
+#include <vector>
+#include "Spire/Ui/TableModel.hpp"
+#include "Spire/Ui/TableModelTransactionLog.hpp"
+#include "Spire/Ui/Ui.hpp"
+
+namespace Spire {
+
+  /** Implements a TableModel that provides a view over a source model by
```
Comment (kranar) [inline Applications/Spire/Include/Spire/Ui/FilteredTableModel.hpp:12]: Put the ```Implements...``` on a new line.

File: Applications/Spire/Source/Ui/FilteredTableModel.cpp [hunk 2]
```diff
@@ -0,0 +1,141 @@
+#include "Spire/Ui/FilteredTableModel.hpp"
+#include <algorithm>
+
+using namespace boost::signals2;
+using namespace Spire;
+
+FilteredTableModel::FilteredTableModel(std::shared_ptr<TableModel> source,
+  std::function<bool(const TableModel&, int)> filter)
+    : m_source(source),
```
Comment (kranar) [inline Applications/Spire/Source/Ui/FilteredTableModel.cpp:9]: Use ```std::move```.

File: Applications/Spire/Source/Ui/FilteredTableModel.cpp [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+      m_filter(filter) {
+  for(auto i = 0; i != static_cast<int>(m_source->get_row_size()); ++i) {
+    if(!m_filter(*m_source.get(), i)) {
```
Comment (kranar) [inline Applications/Spire/Source/Ui/FilteredTableModel.cpp:12]: ```*m_source```

File: Applications/Spire/Source/Ui/FilteredTableModel.cpp [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+  }
+  return std::make_tuple(false, iter);
+}
+
+void FilteredTableModel::on_operation(const Operation& operation) {
+  m_transaction.transact([&] {
+    visit(operation,
+      [&] (const AddOperation& operation) {
+        if(operation.m_index >= m_source->get_row_size() - 1) {
+          if(!m_filter(*m_source.get(), operation.m_index)) {
```
Comment (kranar) [inline Applications/Spire/Source/Ui/FilteredTableModel.cpp:63]: ```*m_source```

File: Applications/Spire/Source/Ui/FilteredTableModel.cpp [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+            m_filtered_data.push_back(operation.m_index);
+            m_transaction.push(AddOperation{
+              static_cast<int>(m_filtered_data.size()) - 1});
+          }
+        } else {
+          auto iter = std::get<1>(find(operation.m_index));
+          std::for_each(iter, m_filtered_data.end(),
+            [] (int& value) { ++value; });
+          if(!m_filter(*m_source.get(), operation.m_index)) {
+            m_transaction.push(AddOperation{m_filtered_data.insert(iter,
```
Comment (kranar) [inline Applications/Spire/Source/Ui/FilteredTableModel.cpp:73]: Cast to ```int```. Generally cast any type that depends on the compiler to an ```int```, so stuff like ```std::size_t```, ```std::ptrdiff```, so on so forth...

File: Applications/Spire/Source/Ui/FilteredTableModel.cpp [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+}
+
+connection FilteredTableModel::connect_operation_signal(
+    const OperationSignal::slot_type& slot) const {
+  return m_transaction.connect_operation_signal(slot);
+}
+
+std::tuple<bool, std::vector<int>::iterator> FilteredTableModel::find(
+    int index) {
+  auto iter = std::lower_bound(m_filtered_data.begin(), m_filtered_data.end(),
```
Comment (kranar) [inline Applications/Spire/Source/Ui/FilteredTableModel.cpp:50]: Rename ```iter``` to ```i```. Feel free to use ```i```, ```j```, ```k``` for iterators.

File: Applications/Spire/Source/Ui/FilteredTableModel.cpp [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+            destination = std::lower_bound(m_filtered_data.begin(), source,
+              operation.m_destination);
+            for(auto iter = source; iter != destination; --iter) {
+              *iter = *(iter - 1) + 1;
+            }
+          }
+          *destination = operation.m_destination;
+          m_transaction.push(MoveOperation{source - m_filtered_data.begin(),
+            destination - m_filtered_data.begin()});
+        } else {
```
Comment (kranar) [inline Applications/Spire/Source/Ui/FilteredTableModel.cpp:99]: else if(operation.m_source < operation.m_destination) {
    } else {
    }

File: Applications/Spire/Source/Ui/FilteredTableModel.cpp [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+        auto iter = std::get<1>(data);
+        if(!m_filter(*m_source.get(), operation.m_row)) {
+          if(std::get<0>(data)) {
+            m_transaction.push(UpdateOperation{iter - m_filtered_data.begin(),
+              operation.m_column});
+          } else {
+            m_transaction.push(AddOperation{m_filtered_data.insert(iter,
+              operation.m_row) - m_filtered_data.begin()});
+          }
+        } else {
```
Comment (kranar) [inline Applications/Spire/Source/Ui/FilteredTableModel.cpp:132]: ```else if(std::get<0>)```

File: Applications/Spire/Source/SpireTester/FilteredTableModelTester.cpp [hunk 9]
```diff
@@ -0,0 +1,430 @@
...
+TEST_SUITE("FilteredTableModel") {
+  TEST_CASE("filter") {
+    auto source = std::make_shared<ArrayTableModel>();
+    source->push({4});
+    source->push({2});
+    source->push({9});
+    source->push({1});
+    auto filtered_model = FilteredTableModel(source,
+      [&] (const TableModel& model, int row) {
+        if(model.get<int>(row, 0) > 3) {
```
Comment (kranar) [inline Applications/Spire/Source/SpireTester/FilteredTableModelTester.cpp:18]: ```return model.get<int>(row, 0) > 3;```

**Review by kranar — CHANGES_REQUESTED** [review]
Looks very good overall.

File: Applications/Spire/Source/Ui/FilteredTableModel.cpp [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+    const OperationSignal::slot_type& slot) const {
+  return m_transaction.connect_operation_signal(slot);
+}
+
+std::tuple<bool, std::vector<int>::iterator> FilteredTableModel::find(
+    int index) {
+  auto iter = std::lower_bound(m_filtered_data.begin(), m_filtered_data.end(),
+    index);
+  if(iter != m_filtered_data.end() && *iter == index) {
+    return std::make_tuple(true, iter);
```
[the remark this answers is not in the document]
Comment (shaobinspire) [reply Applications/Spire/Source/Ui/FilteredTableModel.cpp:53]: Do you mean that std::make_tuple is replaced by a initializer list? like
return {true, iter};

File: Applications/Spire/Source/SpireTester/FilteredTableModelTester.cpp [hunk 11]
```diff
@@ -196,10 +184,7 @@ TEST_SUITE("FilteredTableModel") {
     source->push({3});
     auto filtered_model = FilteredTableModel(source,
       [&] (const TableModel& model, int row) {
```
Comment (kranar) [inline Applications/Spire/Source/SpireTester/FilteredTableModelTester.cpp:186]: Remove the capture by reference if nothing is being captured.

File: Applications/Spire/Source/Ui/FilteredTableModel.cpp [hunk 12]
```diff
@@ -6,10 +6,10 @@ using namespace Spire;
 
 FilteredTableModel::FilteredTableModel(std::shared_ptr<TableModel> source,
   std::function<bool(const TableModel&, int)> filter)
-    : m_source(source),
+    : m_source(std::move(source)),
       m_filter(filter) {
```
Comment (kranar) [inline Applications/Spire/Source/Ui/FilteredTableModel.cpp:10]: Use ```std::move``` for ```filter```.

File: Applications/Spire/Source/Ui/FilteredTableModel.cpp [hunk 13]
```diff
@@ -47,94 +47,91 @@ connection FilteredTableModel::connect_operation_signal(
...
+              static_cast<int>(m_filtered_data.insert(i, operation.m_index) -
+              m_filtered_data.begin())});
           }
         }
       },
       [&] (const MoveOperation& operation) {
-        auto source_row = find(operation.m_source);
-        auto source = std::get<1>(source_row);
-        if(std::get<0>(source_row)) {
+        auto [is_find, source] = find(operation.m_source);
```
Comment (kranar) [inline Applications/Spire/Source/Ui/FilteredTableModel.cpp:80]: ```is_find``` -> ```is_found```

File: Applications/Spire/Source/Ui/FilteredTableModel.cpp [hunk 14]
```diff
@@ -47,94 +47,91 @@ connection FilteredTableModel::connect_operation_signal(
...
         }
       },
       [&] (const RemoveOperation& operation) {
-        auto data = find(operation.m_index);
-        auto iter = std::get<1>(data);
-        std::for_each(iter, m_filtered_data.end(), [] (int& value) { --value; });
-        if(std::get<0>(data)) {
-          auto index = iter - m_filtered_data.begin();
-          m_filtered_data.erase(iter);
+        auto [is_find, i] = find(operation.m_index);
```
Comment (kranar) [inline Applications/Spire/Source/Ui/FilteredTableModel.cpp:111]: ```is_found```

**Review by kranar — CHANGES_REQUESTED** [review]
Some minor fixes remain.

**Pull request merged**